onnx/defs/nn/defs.cc

```diff
@@ -23,6 +23,94 @@ static std::string auto_pad_doc =
 } // namespace ONNX_NAMESPACE
 
 namespace ONNX_NAMESPACE {
+
+void convPoolTypeAndShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+  if (!hasNInputShapes(ctx, 2)) {
+    return;
+  }
+
+  // don't bother with legacy auto_pad for now
+  if (ctx.getAttribute("auto_pad")) {
+    return;
+  }
+
+  size_t n_input_dims = (size_t) (ctx.getInputType(0)->tensor_type().shape().dim_size() - 2);
+
+  std::vector<int64_t> dilations;
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
+    if (dilations.size() != n_input_dims) {
+      return;
+    }
+  } else {
+    for (size_t i = 0; i < n_input_dims; ++i) {
+      dilations.push_back(1);
+    }
+  }
+
+  std::vector<int64_t> kernel_shape;
+  if (getRepeatedAttribute(ctx, "kernel_shape", kernel_shape)) {
+    if (kernel_shape.size() != ctx.getInputType(1)->tensor_type().shape().dim_size() - 2) {
+      return;
+    }
+  } else {
+    for (int i = 2; i < ctx.getInputType(1)->tensor_type().shape().dim_size(); ++i) {
+      kernel_shape.push_back(ctx.getInputType(1)->tensor_type().shape().dim(i).dim_value());
+    }
+  }
+
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "pads", pads)) {
+    if (pads.size() != n_input_dims * 2) {
+      return;
+    }
+  } else {
+    for (size_t i = 0; i < n_input_dims; ++i) {
+      pads.push_back(0);
+      pads.push_back(0);
+    }
+  }
+
+  std::vector<int64_t> strides;
+  if (getRepeatedAttribute(ctx, "strides", strides)) {
+    if (strides.size() != n_input_dims) {
+      return;
+    }
+  } else {
+    for (size_t i = 0; i < n_input_dims; ++i) {
+      strides.push_back(1);
+    }
+  }
+
+  *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape()->add_dim() =
+    ctx.getInputType(0)->tensor_type().shape().dim(0);
+  *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape()->add_dim() =
+    ctx.getInputType(1)->tensor_type().shape().dim(0);
+
+  for (size_t i = 0; i < kernel_shape.size(); ++i) {
+    if (!ctx.getInputType(0)->tensor_type().shape().dim(2 + (int)i).has_dim_value()) {
+      continue;
```

Comment:
Add an (unknown) dim to output shape in this case?

Reply:
++